Beef up pipeline docs (#3555)
tenzir/tenzir: web/docs/pipelines.md

---
sidebar_position: 0
---

# Pipelines

A Tenzir **pipeline** is a chain of **operators** that represents a dataflow.
Operators are the atomic building blocks that produce, transform, or consume
data. Think of them as UNIX or Powershell commands where output from one command
is input to the next:

![Pipeline Chaining](pipeline-chaining.excalidraw.svg)

There exist three types of operators: **sources** that produce data, **sinks**

> that consume data, and transformations that do both:

Reviewer: ```suggestion
that consume data, and **transformations** that do both:
```

![Pipeline Structure](pipeline-structure.excalidraw.svg)

## Polymorphic Operators

Tenzir pipelines make one more distinction: the elements that the operators push
through the pipeline are *typed*. Every operator has an input and an output
type:

![Input and Output Types](operator-pieces.excalidraw.svg)

When composing pipelines out of operators, the type of adjacent operators have
to match. Otherwise the pipeline is malformed. Here's an example pipeline with
matching operators:

![Typed Pipeline](typed-pipeline.excalidraw.svg)

We call any void-to-void operator sequence a **closed pipeline**. Only closed
pipelines can execute. If a pipeline does not have a source and sink, it would
"leak" data. If a pipeline is open, the engine auto-completes a source/sink when

> possible or rejects it.

Reviewer: We should consider mentioning here that the default source/sink depends on how you use Tenzir. In the app, we automatically show the results in the explorer if the sink is missing. On the command-line, we read/write JSON from/to stdin.

Author: Changed to:

> If a pipeline is open, the engine auto-completes a source/sink when
possible and rejects the pipeline otherwise. Auto-completion is context
dependent: on the command line we read JSON from stdin and write it stdout. In
the app we only auto-complete a missing sink with
[`serve`](operators/sinks/serve.md) to display the result in the browser.

Zooming out in the above type table makes the operator types apparent:

![Operator Types](operator-types.excalidraw.svg)

In fact, we can define the operator type as a function of its input and output types:

> ```

Reviewer: Not sure if Docusaurus supports it, but GitHub Markdown recently added support for math blocks by setting the language of the code block to `math`.

```suggestion
```math
```

Author: There is [support for Math equations](https://docusaurus.io/docs/3.0.0-beta.0/markdown-features/math-equations), but it requires setting up KaTeX. So I'd say we do this at some point in the future.

(Input x Output) → {Source, Sink, Transformation}
```

## Multi-Schema Dataflows

Tenzir dataflows are *multi-schema* in that a single pipeline can work
with heterogeneous types of events, each of which have a different schemas.
This allows you, for example, to perform aggregations across multiple events.
Multi-schema dataflows require automatic schema inference at parse time. Tenzir
parsers, such as [`json`](/formats/json) support this out of the box.

This behavior is very different from execution engines that only work with
structured data, where the unit of computation is typically a fixed set of
tables. Schema-less systems, such as document-oriented databases, offer more
simplicity, at the cost of performance. In the spectrum of performance and ease
of use, Tenzir therefore [fills a gap](/why-tenzir):

![Structured vs. Document-Oriented](structured-vs-document-oriented.excalidraw.svg)

:::info Eclectic & Super-structured Data
[Zed](https://amyousterhout.com/papers/zed_cidr23.pdf) has a type system similar
to Tenzir, with the difference that Zed associates types *with every single
value*. Unlike Zed, Tenzir uses a "data frame" abstraction and relies on

> homogeneous Arrow record batches of up to 65,535 rows.

Reviewer: Mostly an FYI, not sure if we need to rewrite this: The limit of $`2^{16}`$ is arbitrarily chosen by us. Some operators can go above it, actually. With the plans for dynamic rebatching as an under-the-hood optimization we actually should go well above it in some cases.

Reviewer: The only limit enforced by Arrow is 2 GiB per array, and that's only because migrating to FlatBuffers64 would be a breaking change. There is basically no limit on record batches themselves, and neither is there a limit on our chunks or table slices. We usually cut at the latest after 64 Ki events, but that's more by convention than a technical necessity.

:::

> If the schema in a pipeline changes, we simply create new batch of events. The

Reviewer: ```suggestion
If the schema in a pipeline changes, we simply create a new batch of events. The
```

worst case for Tenzir is a ordered stream of schema-switching events, with every
event having a new schema than the previous one. That said, even for those data
streams we can efficiently build homogeneous batches when the inter-event order
does not matter significantly. Similar to predicate pushdown, Tenzir operators

> support "orderness pushdown" to signal to upstream operators that the event

Reviewer: ```suggestion
support "ordering pushdown" to signal to upstream operators that the event
```

At least that's how we call it internally.

order only matters intra-schema but not inter-schema. In this case we
transparently demultiplex a heterogeneous stream into *N* homogeneous streams,
each of which yields batches of up to 65k events. The `import` operator is an

> example of such an operator, and it pushes the its orderness upstream so that we

Reviewer: ```suggestion
example of such an operator, and it pushes its ordering expectations upstream so that we
```

can efficiently parse, say, a diverse stream of NDJSON records, such as
Suricata's EVE JSON or Zeek's streaming JSON.

You could call multi-schema dataflows *multiplexed* and there exist dedicated
operators to demultiplex a stream. As of now, this is hard-coded per operator.

> For example, the [`directory`](/connectors/directory) operator demultiplexes a

Reviewer: `directory` is a saver, not an operator. This should probably be rewritten to say `to directory`.

Author: I made this `to directory /tmp/dir write parquet`.

stream of events and writes of the same schema to the same file: with `to
directory write parquet` you can write a set of Parquet files, each of which
have a fixed schema.

The diagram below illustrates the multi-schema aspect of dataflows for schemas
A, B, and C:

![Multi-schema Example](multi-schema-example.excalidraw.svg)

> Some operators only work with exactly one instance per schema internally, such
> as [`parquet`](/formats/parquet), [`feather`](/formats/parquet), or
> [`csv`](/formats/csv). These must be combined with a demultiplexing operator,
> such as the `directory` saver.

Reviewer: All three mentioned examples are formats, not operators. I would recommend rewriting like this:

```suggestion
Some operators only work with exactly one instance per schema internally, such
as the [`write`](/operators/transformations/write) when combined with the
[`parquet`](/formats/parquet), [`feather`](/formats/feather), or
[`csv`](/formats/csv) formats. These formats cannot handle multiple input
schemas at once. A demultiplexing operator like `to directory write parquet`
removes this limitation by writing one file per schema instead.
```

We are having ideas to make this schema (de)multiplexing explicit with a
`per-schema` [operator modifier](/operators/modifier) that you can write in
front of every operator. Similarly, we are going to add union types in the
future, making it possible to convert a heterogeneous stream of structured data

> into a homogeneous one.

Reviewer: Note that such a `fuse` operator would still be blocking, so it's not a good solution to this problem for long-running pipelines.

Author: Right. I think it really depend on the use case here and don't want to go into details.
